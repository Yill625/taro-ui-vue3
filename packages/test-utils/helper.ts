import { mount, shallowMount, DOMWrapper, VueWrapper } from '@vue/test-utils'
import { VNode, Slot } from '@vue/runtime-core'

export const sleep = async (timeout: number): Promise<null> =>
  new Promise((resolve) => setTimeout(resolve, timeout))

export interface Slots {
  default: Array<string | VNode> | Slot
  [x: string]: Array<string | VNode> | Slot
}

export function mountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return mount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export function genMountFn(
  comp: any,
  components?: Record<string, object>,
  useShallowMount?: boolean
) {
  return (props = {}, slots?: Slots) => {
    return useShallowMount
      ? shallowMountFactory(comp, components, props, slots)
      : mountFactory(comp, components, props, slots)
  }
}

export function shallowMountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return shallowMount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export interface Dimension {
  top?: number
  bottom?: number
  left?: number
  right?: number
  width?: number
  height?: number
}

export function genDelayedSelectorSpy(
  utils: any,
  dimensions: Dimension
): jest.SpyInstance {
  return jest
    .spyOn(utils, 'delayQuerySelector')
    .mockImplementation((_, sel: string, delay?: number) => {
      return new Promise((resolve) => {
        resolve([dimensions])
      })
    })
}

export async function triggerTouchEvents(
  el: DOMWrapper<Element>,
  touchstartDetails: any,
  touchendDetails: any
) {
  const startDetails = {
    touches: [touchstartDetails],
    changedTouches: [touchstartDetails]
  }

  const endDetails = {
    touches: [touchendDetails],
    changedTouches: [touchendDetails]
  }

  await el.trigger('touchstart', startDetails)
  await el.trigger('touchmove', endDetails)
  await el.trigger('touchend', endDetails)
}

export function testPropClassAndStyle(
  mountFn: (props?: {}, slots?: Slots | undefined) => VueWrapper<any>
) {

  it.concurrent('should render prop class', async () => {
    const wrapper = mountFn({ class: 'test' })
    expect(
      wrapper.classes()
    ).toContain('test')
  })

  it.concurrent('should render prop style', async () => {
    const wrapper = mountFn({ style: { color: 'red' } })
    expect(
      wrapper.attributes('style')
    ).toContain('color: red;')
  })
}

export function hexToRGBA(hex: string, opacity = 1) {
  hex = hex.replace('#', '')
  const rgb = hex
    .match(new RegExp('(.{' + hex.length / 3 + '})', 'g'))!
    .map((l) => parseInt(hex.length % 2 ? l + l : l, 16))

  const rgba = rgb.concat(isFinite(opacity) ? opacity : 1)

  return opacity === 1
    ? `rgb(${rgb.join(', ')})`
    : `rgba(${rgba.join(', ')})`
}