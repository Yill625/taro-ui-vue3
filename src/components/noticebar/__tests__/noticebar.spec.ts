import { mountFactory, Slots } from '@/tests/helper'
import AtNoticebar from '../index'
import Taro from '@tarojs/taro'

const factory = (
  props = {},
  slots?: Slots
) => {
  return mountFactory(AtNoticebar, undefined, props, slots)
}

describe('AtNoticebar', () => {

  it('should render prop single', () => {
    const wrapper = factory({ single: true })
    expect(
      wrapper
        .find('.at-noticebar--single')
        .exists()
    ).toBeTruthy()
  })

  it('should render prop marquee', () => {
    const wrapper = factory({ marquee: true, single: true })
    expect(
      wrapper
        .find('.at-noticebar--marquee')
        .exists()
    ).toBeTruthy()

    expect(
      wrapper
        .find('.at-noticebar--single')
        .exists()
    ).toBeFalsy()
  })

  it('should render prop speed', () => {
    const wrapper = factory({
      speed: 200,
      marquee: true
    }, {
      default: ['this is a slot content']
    })
    expect(
      wrapper
        .find('.at-noticebar__content-inner')
        .attributes('style')
    ).toBe('animation-duration: 15s;')
  })

  it('should not render moreText in multi-line mode', () => {
    const wrapper = factory({
      moreText: '查看更多',
      showMore: true
    })
    expect(
      wrapper
        .find('.at-noticebar__more')
        .exists()
    ).toBe(false)
  })

  it('should not render moreText in single-line mode', () => {
    const wrapper = factory({
      moreText: '查看更多',
      showMore: false,
      single: true
    })
    expect(
      wrapper
        .find('.at-noticebar__more')
        .exists()
    ).toBe(false)
  })

  it('should render moreText in single-line mode and match snapshot', () => {
    const wrapper = factory({
      moreText: '查看更多',
      showMore: true,
      single: true
    })

    const moreTextEl = wrapper.find('.at-noticebar__more')
    expect(moreTextEl.exists()).toBe(true)
    expect(moreTextEl.element).toMatchSnapshot()
  })

  it('should render close icon and match snapshot', () => {
    const wrapper = factory({ close: true })
    const closeIconEl = wrapper.find('.at-noticebar > .at-noticebar__close')
    expect(closeIconEl.exists()).toBe(true)
    expect(closeIconEl.element).toMatchSnapshot()
  })

  it('should render prop icon and match snapshot', () => {
    const wrapper = factory({ icon: 'volume-plus' })
    const iconEl = wrapper.find('.at-noticebar__content > .at-noticebar__content-icon')
    expect(iconEl.exists()).toBe(true)
    expect(iconEl.element).toMatchSnapshot()
  })

  it('should render slot content', () => {
    const wrapper = factory({}, {
      default: ['this is slot content text']
    })

    expect(
      wrapper
        .find('.at-noticebar__content-inner')
        .text()
    ).toBe('this is slot content text')
  })
})

describe('AtNoticebar Behavior', () => {
  beforeEach(() => {
    jest.mock('@tarojs/taro', () => ({
      ...(jest.requireActual('@tarojs/taro')) as object,
      getEnv: jest.fn(),
    }))
    jest.useFakeTimers()
    jest.spyOn(global, 'setTimeout')
  })

  it('should trigger onClose', async () => {
    const onClose = jest.fn()
    const wrapper = factory({
      close: true,
      onClose: onClose,
    })
    await wrapper.find('.at-noticebar__close').trigger('tap')
    expect(onClose).toBeCalled()
    expect(wrapper.find('.at-noticebar').exists()).toBeFalsy()
  })

  it('should trigger onGotoMore', async () => {
    const onGotoMore = jest.fn()
    const wrapper = factory({
      icon: 'volume-plus',
      single: true,
      showMore: true,
      moreText: '更多内容',
      onGotoMore: onGotoMore
    })
    await wrapper.find('.at-noticebar__more').trigger('tap')
    expect(onGotoMore).toBeCalled()
  })

  it('should initiate h5 animation when mounted if marquee is used', () => {
    const spy = jest
      .spyOn(document, 'querySelector')
      // @ts-ignore
      .mockImplementation((sel: string) => {
        return {
          getBoundingClientRect: jest.fn().mockReturnValueOnce({ width: 30 })
        }
      })

    const wrapper = factory({
      single: true,
      marquee: true,
    })

    jest.runAllTimers()
    expect(setTimeout).toBeCalledTimes(1)
    spy.mockRestore()
  })

  it('should initiate miniapp animation when mounted if marquee is used', () => {
    jest
      .spyOn(Taro, 'getEnv')
      .mockReturnValue(Taro.ENV_TYPE.WEAPP)

    jest.spyOn(Taro, 'createSelectorQuery')
      // @ts-ignore
      .mockImplementation(() => {
        const query = {
          exec: jest.fn()
            .mockImplementation((cb: (rect) => void) => {
              cb([{ width: 30, left: 30 }])
            })
        }
        query['select'] = jest.fn().mockReturnValue({
          boundingClientRect: jest.fn().mockReturnValue(query)
        })

        return query
      })

    const wrapper = factory({
      single: true,
      marquee: true,
    })

    jest.runOnlyPendingTimers()
    jest.advanceTimersByTime(900)
    expect(setTimeout).toBeCalledTimes(4)
  })
})
