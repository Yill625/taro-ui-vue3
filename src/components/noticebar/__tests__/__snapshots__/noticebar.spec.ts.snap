// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AtNoticebar should render close icon and match snapshot 1`] = `
<view
  class="at-noticebar__close"
>
  <text
    class="at-icon at-icon-close"
  />
</view>
`;

exports[`AtNoticebar should render moreText in single-line mode and match snapshot 1`] = `
<view
  class="at-noticebar__more"
>
  <text
    class="text"
  >
    查看更多
  </text>
  <view
    class="at-noticebar__more-icon"
  >
    <text
      class="at-icon at-icon-chevron-right"
    />
  </view>
</view>
`;

exports[`AtNoticebar should render prop icon and match snapshot 1`] = `
<view
  class="at-noticebar__content-icon"
>
  <text
    class="at-icon at-icon-volume-plus"
  />
</view>
`;
